kele.el

```diff
@@ -435,23 +435,109 @@ contents."
   (when bootstrap
     (kele--cache-update cache)))
 
-(defvar kele--context-proxy-ledger nil
-  "An alist mapping contexts to their corresponding proxy processes.
+(defclass kele--proxy-manager ()
+  ((procs
+    :documentation
+    "Alist of context names to proxy processes."
+    :initarg :procs
+    :type list
+    :initform nil)
+   (timers
+    :documentation
+    "Alist of context names to timer objects that terminate and clean up the
+  proxy processes."
+    :initarg :timers
+    :type list
+    :initform nil)
+   (ports
+    :documentation
+    "Alist of context names to the corresponding proxy's port."
+    :initarg :ports
+    :type list
+    :initform nil))
+  "Manage proxy server processes.")
+
+(cl-defmethod proxy-start ((manager kele--proxy-manager)
+                           context
+                           &key port (ephemeral t))
+  "Start a proxy process within MANAGER for CONTEXT at PORT.
+
+If EPHEMERAL is non-nil, the proxy process will be cleaned up
+after a certain amount of time.
+
+If PORT is nil, a random port will be chosen.
+
+Returns the proxy process.
+
+If CONTEXT already has a proxy process active, this function returns the
+existing process *regardless of the value of PORT*."
+  (-if-let ((&alist context proc) (oref manager procs))
+      proc
+    (kele--with-progress (format "Starting proxy server process for `%s'..."
+                                 context)
+      (let* ((selected-port (or port (kele--random-port)))
+             (proc (kele--proxy-process context :port selected-port :wait nil))
+             (cleanup (when ephemeral
+                        (run-with-timer kele-proxy-ttl nil (-partial #'proxy-stop
+                                                                     manager
+                                                                     context)))))
+        (oset manager procs (cons `(,context . ,proc) (oref manager procs)))
+        (oset manager ports (cons `(,context . ,selected-port) (oref manager ports)))
+        (when cleanup
+          (oset manager timers (cons `(,context . ,cleanup) (oref manager timers))))
+        proc))))
+
+(cl-defmethod proxy-get ((manager kele--proxy-manager)
+                         context
+                         &key (wait t))
```

Comment:
As mentioned elsewhere, have this return a container object that includes not only the process but also the port and the cleanup timer.

Might as well `defstruct` it while we're at it.